src/model/styles.rs

```diff
@@ -281,9 +294,100 @@ impl Selector {
         Self::Can(TypeId::of::<T>())
     }
 
+    /// Matches the selector for an introspector.
+    pub fn match_iter<'a>(
+        &'a self,
+        introspector: &'a Introspector,
+    ) -> Box<dyn Iterator<Item = (usize, &'a Content)> + 'a> {
+        self.match_iter_inner(introspector, introspector.all().enumerate())
+    }
+
+    /// Transforms this selector into a [`Selector::Before`] selector.
+    pub fn before(self, location: Location, inclusive: bool) -> Self {
+        Self::Before(Box::new(self), location, inclusive)
+    }
+
+    /// Transforms this selector into a [`Selector::After`] selector.
+    pub fn after(self, location: Location, inclusive: bool) -> Self {
+        Self::After(Box::new(self), location, inclusive)
+    }
+
+    /// Transforms this selector and an iterator of other selectors into a
+    /// [`Selector::Or`] selector.
+    pub fn and(self, others: impl IntoIterator<Item = Self>) -> Self {
+        Self::And(others.into_iter().chain(Some(self)).collect())
```

Comment:
these could automatically optimize nested ands

Reply:
This will need to be done in a separate optimization PR, this is not the only place where we might eke out extra performance, but I feel like this PR is already big enough